src/pathways/services/pathwayUtils.service.ts

```diff
@@ -0,0 +1,151 @@
+            .flatten()
+            .sortBy()
+            .sortedUniq()
+            .value();
+    }
+
+    /**
+     * Groups reactions by metabolite name
+     * @param reactions 
+     * @param metabolites 
+     */
+    public reactionsByMetabolite(reactions, metabolites) {
+        return _.zipObject(metabolites, _.map(metabolites, (m) => {
+            return _.filter(reactions, (r)=> _.has(r.metabolites, m));
+        }));
+    }
+
+    /**
+     * Returns metabolites with 2 or less reactions
+     * @param reactionsByMetabolite 
+     */
+    public obtainCofactors(reactionsByMetabolite) {
+        return _(reactionsByMetabolite)
+            .pickBy(reacions => reacions.length > 2)
+            .keys()
+            .value();
+    }
+
+    public sortedReactionsProducts(reactions, product) {
+        // First try to replace the backend functionality
+        // const metabolites = _.reject(this.reactionsMetabolites(reactions), 
+        //                              x => this.DEFAULT_COFACTORS.has(x));
+        const metabolites = this.reactionsMetabolites(reactions);
+
+        // const metabolitesByName = _.keyBy(metabolites, 'name');
+        const reactionsByMetabolite = this.reactionsByMetabolite(reactions, metabolites);
+
+        // will be removed.
+        const customCofactors = new Set(this.obtainCofactors(reactionsByMetabolite));
+        // Why not using model.metabolites?
+
+        const graph = new jsnx.DiGraph();
+        const [startReaction] = reactionsByMetabolite[product];
+        graph.addNode(startReaction.id);
+        const cofactors = this.setUnion(this.DEFAULT_COFACTORS, customCofactors);
+        const queue = this.notSecondaryMetabolites(startReaction.metabolites, product, cofactors);
+
+        const primaryNodes = {};
+        primaryNodes[startReaction.id] = product;
+        let traversalState = {
+            graph: graph,
+            // Rename to visited reactions
+            reactions: [startReaction.name],
+            primaryNodes: primaryNodes,
+            reactionsByMetabolite,
+            cofactors: cofactors,
+        };
+        _.forEach(queue, (metabolite) => {
+            traversalState = this.fillReactionsGraph(startReaction, traversalState, metabolite);
+        });
+        const sortedReactions = jsnx.topologicalSort(traversalState.graph);
+        return {
+            reactions: sortedReactions,
+            primaryNodes: _.at(traversalState.primaryNodes, sortedReactions),
+        };
+    }
+
+    /**
+     * Returns all metabolites from the opposite side of the equation
+     * @param metabolites 
+     * @param metabolite 
+     */
+    public oppositeMetabolites(metabolites, metabolite) {
+        const metaboliteCoef = metabolites[metabolite];
+        return _(metabolites)
+            .pickBy((v) => v * metaboliteCoef < 0)
+            .keys()
+            .value();
+    }
+
+    public notSecondaryMetabolites(metabolites, metabolite, cofactors) {
```

Comment:
The implementation here seems substantially different from the one that is being replaced. In the original implementation, cofactors are specified on the object. It would make sense to refactor this code into a very a PathwayGraph class similar to the Python implementation and transform any helper functions into module level functions that are initialized without the use of a service. 

A service only makes sense for facilitating dependency injection and in this case I don't think much of or any of the code having this as a dependency could be mocked up to not use these utility functions. If there is any state, such a PathwayGraph initialized with a configurable set of cofactors, then that can be injected through a service, but simple utility functions do not benefit from being exposed through a service.